lock/etcd.go

```diff
@@ -114,10 +129,6 @@ func (c *EtcdLockClient) Set(sem *Semaphore) error {
 		return err
 	}
 
-	setopts := &client.SetOptions{
-		PrevIndex: sem.Index,
-	}
-
-	_, err = c.keyapi.Set(context.Background(), c.keypath, string(b), setopts)
+	_, err = c.keyapi.Put(context.Background(), c.keypath, string(b))
```

Comment:
I think this needs to be turned into a transaction too, because the old code had a SetOption to apply the changes if the previous index is a value known to us. I think that with v3 it could be something like:
```go
response, err = c.keyapi.Txn(context.Background()).If
	client.Compare(client.Version(c.keypath), "=", sem.Index),
).Then(
	client.OpPut(c.keypath, string(payload)),
).Commit()
if err != nil {
	return err
}
// If it's true, it means that the "then" branch was taken, false - "else" branch was taken.
if !response.Succeeded {
	return fmt.Errorf("failed to set the semaphore - it got updated in the meantime")
}
return nil
```

Also, maybe the `Index` field in the `Semaphore` struct should be renamed to `Version`?

This change will probably allow us to get rid of the `Set` method in the `KeyAPI` interface.

Reply:
I agree - using `Txn` instead of `Put` will provide more controls on the insertion. The `s/Index/Version/` makes sense too.

Reply:
It seems it would be nice to have some congestion tests around that to make sure things behave as expected in distributed environment.

Reply:
@invidian I don't know what are congestion tests - do you have some example to provide ?

Reply:
I meant, if more than one client initializes etc at the same time. Maybe https://github.com/golang/mock could be useful here to simulate the races?

Reply:
I do agree that we should cover this section - but I also think that `etcd` provides the right tool: [`concurrency.Mutex`](https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency) to ensure that we don't have this kind of race conditions.

> Package concurrency implements concurrency operations on top of etcd such as distributed locks, barriers, and elections.

When I started to upgrade to `v3` I knew that the changes will be quite something so I tried to mitigate the migration impact. That's why I did not used the [`concurrency.Mutex`](https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency) as mentioned in the PR description:
> did not use the client/v3/concurrency. in order to not break the current locksmithctl semaphore logic

But now, I feel we are actually redeveloping this [`concurrency.Mutex`](https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency) through `Txn` and racing tests - so it might actually be the correct choice to go ahead with the `Mutex`. What do you think ? :)

Reply:
> I do agree that we should cover this section - but I also think that `etcd` provides the right tool: [`concurrency.Mutex`](https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency) to ensure that we don't have this kind of race conditions.
> 
> > Package concurrency implements concurrency operations on top of etcd such as distributed locks, barriers, and elections.
> 
> When I started to upgrade to `v3` I knew that the changes will be quite something so I tried to mitigate the migration impact. That's why I did not used the [`concurrency.Mutex`](https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency) as mentioned in the PR description:
> 
> > did not use the client/v3/concurrency. in order to not break the current locksmithctl semaphore logic
> 
> But now, I feel we are actually redeveloping this [`concurrency.Mutex`](https://pkg.go.dev/go.etcd.io/etcd/client/v3/concurrency) through `Txn` and racing tests - so it might actually be the correct choice to go ahead with the `Mutex`. What do you think ? :)

Up to you. :) On one hand it would probably make the code easier to understand and the tests would be easier to write (maybe). On the other hand, it introduces more back-and-forth communication with etcdserver. Not sure if the latter is a big problem, so using the mutex may be a good idea.

Reply:
Mm, just quickly read the documentation - it seems we can't really use the `concurrency.Mutex` because it does not allow to use custom semaphores. It relies on the `key` to lock/unlock - so we need to think about how we could authorize _many_ instances ( > 1) to reboot - in other terms, how we could implement the `set-max` command using `concurrency.Mutex`. (see also: https://github.com/etcd-io/etcd/issues/12490)